katharsis-client/src/main/java/io/katharsis/client/internal/BaseResponseDeserializer.java

```diff
@@ -0,0 +1,198 @@
+
+			return collectionResponse;
+		} else {
+			JsonApiResponse apiResponse = new JsonApiResponse();
+
+			if (dataBodies.resources.size() == 1) {
+				apiResponse.setEntity(dataBodies.resources.get(0));
+			} else {
+				apiResponse.setEntity(null);
+			}
+
+			ResourceResponseContext resourceResponse = new ResourceResponseContext(apiResponse, -1);
+			return resourceResponse;
+		}
+	}
+	
+	class ClientResourceUpsert extends ResourceUpsert{
+		
+		private HashMap<Object, Object> resourceMap = new HashMap<Object,Object>();
+
+		public ClientResourceUpsert(ResourceRegistry resourceRegistry, TypeParser typeParser,
+				ObjectMapper objectMapper) {
+			super(resourceRegistry, typeParser, objectMapper);
+		}
+		
+		public String getUID(DataBody body){
+			return body.getType() + "#" + body.getId();
+		}
+
+		public void setRelations(ResourceBodies dataBodies) {
+			for(DataBody body : dataBodies.dataBodies){
+				String uid = getUID(body);
+				Object resource = resourceMap.get(uid);
+				
+				RegistryEntry registryEntry = resourceRegistry.getEntry(body.getType());
+				QueryParams queryParams = null;
+				RepositoryMethodParameterProvider parameterProvider = null;
+				setRelations(resource, registryEntry, body, queryParams, parameterProvider);
+			}			
+		}
+			
+		public void allocateResources(ResourceBodies dataBodies) {
+			for(DataBody body : dataBodies.dataBodies){
+				
+				RegistryEntry<?> registryEntry = resourceRegistry.getEntry(body.getType());
+				ResourceInformation resourceInformation = registryEntry.getResourceInformation();
+				
+				Object resource = newResource(resourceInformation, body);
+				setId(body, resource, resourceInformation);
+				setAttributes(body, resource, resourceInformation);
+				dataBodies.resources.add(resource);
+				
+				String uid = getUID(body);
+				resourceMap.put(uid, resource);
+			}			
+		}
+
+		@Override
+		public boolean isAcceptable(JsonPath jsonPath, String requestType) {
+			throw new IllegalStateException();
+		}
+
+		@Override
+		public BaseResponseContext handle(JsonPath jsonPath, QueryParams queryParams,
+				RepositoryMethodParameterProvider parameterProvider, RequestBody requestBody) {
+			throw new IllegalStateException();
+		}
+		
+		public ResourceBodies parse(JsonNode node, JsonParser jp) throws JsonProcessingException {
+			ResourceBodies bodies = new ResourceBodies();
+			if (node != null) {
+				if (node.isArray()) {
+					Iterator<JsonNode> nodeIterator = node.iterator();
+					while (nodeIterator.hasNext()) {
+						DataBody newLinkage = jp.getCodec().treeToValue(nodeIterator.next(), ClientDataBody.class);
+						bodies.dataBodies.add(newLinkage);
+					}
+					bodies.isCollection = true;
+				} else if (node.isObject()) {
+					bodies.dataBodies.add(jp.getCodec().treeToValue(node, ClientDataBody.class));
+				} else if (node.isNull()) {
+				} else {
+					throw new RuntimeException("data field has wrong type: " + node.toString());
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS00112)